_posts/2024-09-10-building-browser.md

```diff
@@ -0,0 +1,175 @@
+We will also define more specific webview types to satisfy different purposes in the future.
+
+## Multi-window
+
+The other prominent feature after multi-view is the ability to support multiple windows.
+This one wasn’t planned at first, but because it affects too many components, we ended up resolving them together from the ground up.
+
+Servo uses [WebRender](https://github.com/servo/webrender), based on OpenGL, to render its layout.
+To support multiple windows, we need to support multiple OpenGL surfaces.
+One approach would be to create separate OpenGL contexts for each window.
+But since our implementations of WebGL, WebGPU, and WebXR are all tied to a single WebRender instance, which in turn only supports a single OpenGL context for now, we chose to use a single context with multiple surfaces.
+This alternative approach could potentially use less memory and spawn fewer threads.
+For more details, see [this series of blog posts](https://wusyong.github.io/posts/verso-compositor-part3/) by [@wusyong](https://github.com/wusyong).
+
+<figure class="_fig"><a href="{{ '/img/blog/multiwindow.png' | url }}"><img src="{{ '/img/blog/multiwindow.png' | url }}"
+    alt="Verso displaying two windows"></a>
+<figcaption>Figure 2: Verso creates two separate windows with the same OpenGL context.</figcaption></figure>
+
+There is still room for improvement.
+For example, WebRender currently [only supports rendering a single “document”](https://github.com/servo/webrender/blob/0.64/webrender/src/renderer/mod.rs#L1228).
+Unless we create multiple WebRender instances, like Firefox does, we have one WebRender document that has to constantly update all of its display lists to show on all of our windows.
+This could potentially lead to race conditions where a webview may draw to the wrong window for a split second.
+
+There are also different OpenGL versions across multiple platforms, which can be challenging to configure and link.
+Verso is experimenting with using [Glutin](https://docs.rs/glutin/latest/glutin/) for better configuration and attempting to get closer to the general Rust ecosystem.
+
+## What’s next?
+
+With multi-view and multi-window support as the fundamental building blocks, we could create more UI elements to keep pushing the envelope of our browser and embedding research.
+At the same time, Servo is a huge project, with many potential improvements still to come, so we want to reflect on our progress and decide on our priorities.
+Here are some directions that are worth pursuing.
+
+### Benchmarking and metrics
+
+We want to gather the strength of the community to help us clear out the statistics of supported CSS properties and web APIs in Servo by [popularity order](https://chromestatus.com/metrics/css/popularity) and benchmark results such as jetstream2 and speedometer3.
+[@sagudev](https://github.com/sagudev) already started [a subset of speedometer3](https://speedometer-9x8.pages.dev/) to experiment.
```

Comment:
```suggestion
[@sagudev](https://github.com/sagudev) already started [a subset of speedometer3](https://servospeedometer.netlify.app/) to experiment.
```
as this is the url we use in mach: https://github.com/servo/servo/blob/6d6cd0f2dc395dc8dd50a0df1b54bee48d9b1ae6/python/servo/testing_commands.py#L530